File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.java
import java.util.*;
import java.io.*;


class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N;
    static int[] nums = new int[10];
    static boolean[] visited = new boolean[10];
    static int maximum = -1000;
    public static void recur(int step, int result, int prev) {

        if (step == N) {
            maximum = Math.max(result, maximum);
            return;
        }



        for (int i = 0; i < N; i++) {
            if (visited[i]) continue;
            visited[i] = true;
            recur(step+1, (step > 0 ? result + Math.abs(nums[i] - prev) : result), nums[i]);
            visited[i] = false;
        }
    }

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());

        for (int i = 0; i < N; i++)
            nums[i] = Integer.parseInt(st.nextToken());

        recur(0, 0, 0);
        System.out.println(maximum);
    }
}

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.c
#include <stdio.h>
typedef struct _state {
    int step, result, prev, mask;
} State;

State stack[100];
int top = 0;
int nums[10];
int N;

int abs(int x) {
    return (x > 0) ? x : -x;
}

int main(void) {
    scanf("%d", &N);

    for (int i = 0; i < N; i++) scanf("%d", nums + i);

    // State s = 
    stack[top++] = (State){0, 0, 0, 0};

    int maximum = -1;
    while (top != 0) {
        State s = stack[--top];

        if (s.step == N && s.result > maximum) 
            maximum = s.result;
        else {
            for (int i = 0; i < N; i++) {
                if ((s.mask >> i) & 1) continue;
                
                stack[top++] = (State){
                    s.step + 1,
                    (s.step > 0 ? 
                        s.result + abs(nums[i] - nums[s.prev]) :
                        s.result
                    ),
                    i,
                    s.mask | (1 << i)
                };

            }
        }
    }

    printf("%d\n", maximum);
    return 0;
}